pkg/manager/member/pump_scaler.go

```diff
@@ -0,0 +1,282 @@
+			if !strings.Contains(cmd, "-advertise-addr") {
+				continue
+			}
+
+			// example:
+			// ...
+			// /pump \
+			// -pd-urls=http://basic-pd:2379 \
+			// -L=info \
+			// -advertise-addr=`echo ${HOSTNAME}`.basic-pump:8250 \
+			// -config=/etc/pump/pump.toml \
+			// -data-dir=/data \
+			// -log-file=
+			// ...
+			for _, str := range strings.Split(cmd, "\n") {
+				if !strings.Contains(str, "-advertise-addr") {
+					continue
+				}
+
+				str = strings.Replace(str, "`echo ${HOSTNAME}`", pod.Name, -1)
+				str = strings.Replace(str, "-advertise-addr=", "", -1)
+				str = strings.TrimRight(str, "\\")
+				str = strings.TrimSpace(str)
+				return str
+			}
+		}
+	}
+
+	return ""
+}
+
+func (s *pumpScaler) ScaleIn(meta metav1.Object, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	ns := meta.GetNamespace()
+	tcName := meta.GetName()
+
+	// we can only remove one member at a time when scaling in
+	_, ordinal, replicas, deleteSlots := scaleOne(oldSet, newSet)
+	resetReplicas(newSet, oldSet)
+
+	// ref: https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#deployment-and-scaling-guarantees
+	// Before a scaling operation is applied to a Pod, all of its predecessors must be Running and Ready.
+	//
+	// after offline the pump, the pump will be not ready, so we only offline the next one after the successors pod are deleted.
+	if *oldSet.Spec.Replicas != oldSet.Status.Replicas {
+		return controller.RequeueErrorf("Wait for statefulset to be synced before scale in one more replica, desired: %d, running: %d", *oldSet.Spec.Replicas, oldSet.Status.Replicas)
+	}
+
+	klog.Infof("scaling in pump statefulset %s/%s, ordinal: %d (replicas: %d, delete slots: %v)", oldSet.Namespace, oldSet.Name, ordinal, replicas, deleteSlots.List())
+	// We need remove member from cluster before reducing statefulset replicas
+	var podName string
+
+	switch meta.(type) {
+	case *v1alpha1.TidbCluster:
+		podName = ordinalPodName(v1alpha1.PumpMemberType, tcName, ordinal)
+	default:
+		return fmt.Errorf("pumpScaler.ScaleIn: failed to convert cluster %s/%s", meta.GetNamespace(), meta.GetName())
+	}
+	pod, err := s.deps.PodLister.Pods(ns).Get(podName)
+	if err != nil {
+		return fmt.Errorf("pumpScaler.ScaleIn: failed to get pods %s for cluster %s/%s, error: %s", podName, ns, tcName, err)
+	}
+
+	tc, _ := meta.(*v1alpha1.TidbCluster)
+
+	client, err := buildBinlogClient(tc, s.deps.PDControl)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	// Since the advertise address may no contains the namespace
+	// and operator do not run in the same namespace with tidb-cluster,
+	// we can not use this advertise address to access pump.
+	// so will add the namespace part to the address if need.
+	client.HookAddr = func(addr string) string {
+		host, port, err := net.SplitHostPort(addr)
+		if err != nil {
+			return addr
+		}
+
+		suffix := "." + ns + ".svc"
+
+		if strings.Contains(addr, suffix) {
+			return addr
+		}
+
+		host += suffix
```

Comment:
No need to include `.svc` in the host, which will cause more DNS queries.

Reply:
check the comment
```
      // Since the advertise address may no contains the namespace
      // and operator do not run in the same namespace with tidb-cluster,
      // we can not use this advertise address to access pump.
      // so will add the namespace part to the address if need.
```

Reply:
I mean only need to add `ns` but no need to add `ns.svc` here.